pkg/kubelet/clustertrustbundle/clustertrustbundle_manager.go

```diff
@@ -0,0 +1,241 @@
+}
+
+type InformerManager struct {
+	ctbInformer cache.SharedIndexInformer
+	ctbLister   certlistersv1alpha1.ClusterTrustBundleLister
+
+	normalizationCache *lrucache.LRUExpireCache
+	cacheTTL           time.Duration
+}
+
+var _ Manager = (*InformerManager)(nil)
+
+func NewInformerManager(bundles certinformersv1alpha1.ClusterTrustBundleInformer, cacheSize int, cacheTTL time.Duration) (*InformerManager, error) {
+	// We need to call Informer() before calling start on the shared informer
+	// factory, or the informer won't be registered to be started.
+	m := &InformerManager{
+		ctbInformer:        bundles.Informer(),
+		ctbLister:          bundles.Lister(),
+		normalizationCache: lrucache.NewLRUExpireCache(cacheSize),
+		cacheTTL:           cacheTTL,
+	}
+
+	// Have the informer bust cache entries when it sees updates that could
+	// apply to them.
+	_, err := m.ctbInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			ctb, ok := obj.(*certificatesv1alpha1.ClusterTrustBundle)
+			if !ok {
+				return
+			}
+			klog.InfoS("Dropping all cache entries for signer", "signerName", ctb.Spec.SignerName)
+			m.dropCacheFor(ctb)
+		},
+		UpdateFunc: func(old, new any) {
+			ctb, ok := new.(*certificatesv1alpha1.ClusterTrustBundle)
+			if !ok {
+				return
+			}
+			klog.InfoS("Dropping cache for ClusterTrustBundle", "signerName", ctb.Spec.SignerName)
+			m.dropCacheFor(new.(*certificatesv1alpha1.ClusterTrustBundle))
+		},
+		DeleteFunc: func(obj any) {
+			ctb, ok := obj.(*certificatesv1alpha1.ClusterTrustBundle)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					return
+				}
+				ctb, ok = tombstone.Obj.(*certificatesv1alpha1.ClusterTrustBundle)
+				if !ok {
+					return
+				}
+			}
+			klog.InfoS("Dropping cache for ClusterTrustBundle", "signerName", ctb.Spec.SignerName)
+			m.dropCacheFor(ctb)
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("while registering event handler on informer: %w", err)
+	}
+
+	return m, nil
+}
+
+func (m *InformerManager) dropCacheFor(ctb *certificatesv1alpha1.ClusterTrustBundle) {
+	if ctb.Spec.SignerName != "" {
+		m.normalizationCache.RemoveAll(func(key any) bool {
+			return key.(cacheKeyType).signerName == ctb.Spec.SignerName
+		})
+	} else {
+		m.normalizationCache.RemoveAll(func(key any) bool {
+			return key.(cacheKeyType).ctbName == ctb.ObjectMeta.Name
+		})
+	}
+}
+
+func (m *InformerManager) GetTrustAnchorsByName(name string, allowMissing bool) ([]byte, error) {
+	if !m.ctbInformer.HasSynced() {
+		return nil, fmt.Errorf("ClusterTrustBundle informer has not yet synced")
+	}
+
+	ctb, err := m.ctbLister.Get(name)
+	if k8serrors.IsNotFound(err) && allowMissing {
+		return []byte{}, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("while getting ClusterTrustBundle: %w", err)
+	}
+
+	cacheKey := cacheKeyType{ctbName: name}
+
+	if cachedAnchors, ok := m.normalizationCache.Get(cacheKey); ok {
+		return cachedAnchors.([]byte), nil
+	}
```

Comment:
This should be before the call to `m.ctbLister.Get(name)` right?

Reply:
Right, done.